### PR TITLE
Made endpoints in links deterministic

### LINKED ISSUE #221: Deterministic instantiation of endpoint_a and endpoint_b in Link
Currently, the `Link.id` property utilizes a deterministic approach to store and calculate the link ID, based on [ascending sort of the DPIDs and ports](https://github.com/kytos-ng/kytos/blob/f05059fd52fd5ec64e3bfe3ce62bf87cb6e50ec3/kytos/core/link.py#L75-L86).

On the other hand, the attributes `Link.endpoint_a` and `Link.endpoint_b` are not sorted. This leads to situations where, depending on which event was handled first by of_lldp, the same link can be stored in different order in the endpoints:

```
## start kytos and do some updates in the topology...

root@7767801619fb:/# curl -s http://127.0.0.1:8181/api/kytos/topology/v3/ > topology1.json

## eventually restart kytos and wait topology to be loaded

root@7767801619fb:/# curl -s http://127.0.0.1:8181/api/kytos/topology/v3/ > topology2.json

root@7767801619fb:/# diff -Naur <(jq --sort-keys . topology1.json) <(jq --sort-keys . topology2.json)
--- /dev/fd/63	2022-05-20 20:34:16.829761741 +0000
+++ /dev/fd/62	2022-05-20 20:34:16.829761741 +0000
@@ -7,38 +7,38 @@
         "endpoint_a": {
           "active": true,
           "enabled": true,
-          "id": "00:00:00:00:00:00:00:18:11",
+          "id": "00:00:00:00:00:00:00:11:11",
           "link": "1e7e531ae50f419b1d35d311a66351d997ea567348b691d8dccdc639e1f8340a",
           "lldp": true,
-          "mac": "aa:72:80:89:7a:af",
+          "mac": "26:6e:bc:27:57:5b",
           "metadata": {
             "mtu": 9000,
-            "port_name": "Interface-18:11"
+            "port_name": "Interface-11:11"
           },
-          "name": "Ampath4-eth11",
+          "name": "Ampath1-eth11",
           "nni": true,
           "port_number": 11,
           "speed": 1250000000,
-          "switch": "00:00:00:00:00:00:00:18",
+          "switch": "00:00:00:00:00:00:00:11",
           "type": "interface",
           "uni": false
         },
         "endpoint_b": {
           "active": true,
           "enabled": true,
-          "id": "00:00:00:00:00:00:00:11:11",
+          "id": "00:00:00:00:00:00:00:18:11",
           "link": "1e7e531ae50f419b1d35d311a66351d997ea567348b691d8dccdc639e1f8340a",
           "lldp": true,
-          "mac": "26:6e:bc:27:57:5b",
+          "mac": "aa:72:80:89:7a:af",
           "metadata": {
             "mtu": 9000,
-            "port_name": "Interface-11:11"
+            "port_name": "Interface-18:11"
           },
-          "name": "Ampath1-eth11",
+          "name": "Ampath4-eth11",
           "nni": true,
           "port_number": 11,
           "speed": 1250000000,
-          "switch": "00:00:00:00:00:00:00:11",
+          "switch": "00:00:00:00:00:00:00:18",
           "type": "interface",
           "uni": false
         },
@@ -46,6 +46,7 @@
         "metadata": {
           "availability": 100,
           "bandwidth": 100,
+          "last_status_change": 1653078485.537819,
           "last_status_is_active": true,
           "latency": 1,
           "link_name": "Ampath4-eth11--Ampath1-eth11",
@@ -59,38 +60,38 @@
         "endpoint_a": {
           "active": true,
           "enabled": true,
-          "id": "00:00:00:00:00:00:00:15:6",
+          "id": "00:00:00:00:00:00:00:16:6",
           "link": "21c046f5eb9fbf577701974e207c2fd2ccd8cc4b91fae77ad396924b79c48483",
           "lldp": true,
-          "mac": "72:a1:47:7f:b4:15",
+          "mac": "76:18:c6:24:64:06",
           "metadata": {
             "mtu": 9000,
-            "port_name": "Interface-15:6"
+            "port_name": "Interface-16:6"
           },
-          "name": "AL2-eth6",
+          "name": "AL3-eth6",
           "nni": true,
           "port_number": 6,
           "speed": 1250000000,
-          "switch": "00:00:00:00:00:00:00:15",
+          "switch": "00:00:00:00:00:00:00:16",
           "type": "interface",
           "uni": false
         },
         "endpoint_b": {
           "active": true,
           "enabled": true,
-          "id": "00:00:00:00:00:00:00:16:6",
+          "id": "00:00:00:00:00:00:00:15:6",
           "link": "21c046f5eb9fbf577701974e207c2fd2ccd8cc4b91fae77ad396924b79c48483",
           "lldp": true,
-          "mac": "76:18:c6:24:64:06",
+          "mac": "72:a1:47:7f:b4:15",
           "metadata": {
             "mtu": 9000,
-            "port_name": "Interface-16:6"
+            "port_name": "Interface-15:6"
           },
-          "name": "AL3-eth6",
+          "name": "AL2-eth6",
           "nni": true,
           "port_number": 6,
           "speed": 1250000000,
-          "switch": "00:00:00:00:00:00:00:16",
+          "switch": "00:00:00:00:00:00:00:15",
           "type": "interface",
           "uni": false
         },
@@ -98,6 +99,7 @@
         "metadata": {
           "availability": 100,
           "bandwidth": 100,
+          "last_status_change": 1653078485.5223212,
           "last_status_is_active": true,
           "latency": 1,
           "link_name": "AL2-eth6--AL3-eth6",
...
```

Maybe we could sort the endpoint when instantiating a Link, instead of at the time of `Link.id` calculation.

Since this has no impact other than facilitating visualization and convenience, the criticality of this issue is very low.

### PR DESCRIPTION
**Pull request by Ktmi**
Closes #221

### Summary

Makes it so that the order of endpoints in a `Link` is deterministic. Uses `LinkID` in order to determine order of endpoints. `LinkID` also updated to have endpoints be ordered in the `interfaces` attribute.

### Local Tests

Ran `tox` and all existing tests are passing.

### End-to-End Tests

Have not yet run e2e tests with this change yet.

### CODE REVIEW

**Review by viniarck — COMMENTED** [review]
@Ktmi, nicely done, looks great to me. 

Can you update the changelog? Also, can you run e2e tests with this change (I don't expect surprises there, but it's good to explore that dynamic LLDP created links are all set)?

**Comment by Ktmi** [comment]
> @Ktmi, nicely done, looks great to me.
> 
> Can you update the changelog? Also, can you run e2e tests with this change (I don't expect surprises there, but it's good to explore that dynamic LLDP created links are all set)?

Can confirm that lldp creates the appropriate links. However, I found that several e2e tests assume the previous endpoint ordering behavior. Here's the log from running the tests: https://gist.github.com/Ktmi/aaa0512a89fa5176422dd07e4e3d196a

**Comment by Ktmi** [comment]
Looking a bit deeper at the failing tests, it seems that `mef_eline` checks that a series of links in the backup path also form a continuous series of switches.

```
____ TestE2EMefEline.test_005_on_primary_path_fail_should_migrate_to_backup ____

self = <tests.test_e2e_11_mef_eline.TestE2EMefEline object at 0x7fd917a3ff10>

    def test_005_on_primary_path_fail_should_migrate_to_backup(self):
    
        # TODO Check for false positives between uni_a switch 1 and uni_z switch 3 instead of switch 2
        """ When the primary_path is down and backup_path exists and is UP
            the circuit will change from primary_path to backup_path. """
        payload = {
            "name": "my evc1",
            "enabled": True,
            "uni_a": {
                "interface_id": "00:00:00:00:00:00:00:01:1",
                "tag": {
                    "tag_type": 1,
                    "value": 101
                }
            },
            "uni_z": {
                "interface_id": "00:00:00:00:00:00:00:02:1",
                "tag": {
                    "tag_type": 1,
                    "value": 101
                }
            },
            "primary_path": [
                {"endpoint_a": {"id": "00:00:00:00:00:00:00:01:3"},
                 "endpoint_b": {"id": "00:00:00:00:00:00:00:02:3"}}
            ],
            "backup_path": [
                {"endpoint_a": {"id": "00:00:00:00:00:00:00:01:4"},
                 "endpoint_b": {"id": "00:00:00:00:00:00:00:04:4"}},
                {"endpoint_a": {"id": "00:00:00:00:00:00:00:04:3"},
                 "endpoint_b": {"id": "00:00:00:00:00:00:00:03:4"}},
                {"endpoint_a": {"id": "00:00:00:00:00:00:00:03:1"},
                 "endpoint_b": {"id": "00:00:00:00:00:00:00:02:4"}}
            ]
        }
    
        api_url = KYTOS_API + '/mef_eline/v2/evc/'
        response = requests.post(api_url, data=json.dumps(payload), headers={'Content-type': 'application/json'})
>       assert response.status_code == 201, response.text
E       AssertionError: {"code":400,"description":"backup_path is not valid: Interface('s3-eth4', 4, Switch('00:00:00:00:00:00:00:03')) switch is different from previous.","name":"Bad Request"}
E         
E       assert 400 == 201
E        +  where 400 = <Response [400]>.status_code

tests/test_e2e_11_mef_eline.py:84: AssertionError
```

**Comment by Ktmi** [comment]
I was thinking, maybe I could modify `mef_eline` to not care about the order of endpoints, but the problem is that there is implied directionality, that I think may potentially matter in the case of looping back to the same switch. @viniarck What do you think?

**Comment by viniarck** [comment]
> I was thinking, maybe I could modify `mef_eline` to not care about the order of endpoints, but the problem is that there is implied directionality, that I think may potentially matter in the case of looping back to the same switch. @viniarck What do you think?

@Ktmi, I was looking into `mef_eline` code. Since `Path` inherents from a `list` the directionality is based on this ordered sequence, so this loop example shouldn't be an issue in terms of validation or detecting it. I think if you were to adapt the [`is_valid` method](https://github.com/kytos-ng/mef_eline/blob/master/models/path.py#L47) to avoid assuming that `endpoint_a` would always be the an interface of the next switch that should work with the current changes I suppose. 

Since in some cases prior to this change, a Link's endpoint wasn't deterministic, this problem that you mentioned already existed in certain circumstances, since whoever is consuming the api would need to know upfront which `endpoint_a` a Link ended up being constructed with and having to send the expected request payload. 

One additional implication to consider is: do we want to accept either `primary_path` and `secondary_path` list payload with `endpoint_a` and `endpoint_b` keys ordered or in any order? I think we could and should without too much extra work or complexity, since ultimately, the deterministic `Link` will be built and serialized and stored in the DB as well. The only potential surprise would in the output of the serialized circuits that the order of `endpoint_a` or `endpoint_b` might not be the same as the user specified, which is OK, as long as we also document in the changelog explaining this.

cc'ing also @italovalcy and @ajoaoff to since they originally were involved in this part validating `mef_eline` static paths if they want to contribute to the discussion.

**Comment by viniarck** [comment]
> Reviving this PR, I adjusted `mef_eline` to these changes PR [#399](https://github.com/kytos-ng/mef_eline/pull/339) and small changes to e2e tests were needed too PR [#244](https://github.com/kytos-ng/kytos-end-to-end-tests/pull/244)

@Alopalao thanks for helping out with this, I ended up hitting a new corner case though with `mef_eline` that I described on https://github.com/kytos-ng/mef_eline/pull/339#discussion_r1237289869

**Review by viniarck — APPROVED** [review]
Leaving this PR pre-approved, @Alopalao could you also update the changelog? Nicely done on how you found a solution on `mef_eline`. Congrats, guys, great collaboration.

**Comment by Alopalao** [comment]
Ran modified e2e tests and with latest changes
```
Ran 'hello' command on MongoDB successfully. It's ready!
+ python3 -m pytest tests/test_e2e_14_mef_eline.py tests/test_e2e_10_mef_eline.py tests/test_e2e_11_mef_eline.py tests/test_e2e_12_mef_eline.py tests/test_e2e_13_mef_eline.py tests/test_e2e_15_mef_eline.py --reruns 2 -r fEr
============================= test session starts ==============================
platform linux -- Python 3.9.2, pytest-7.2.0, pluggy-1.0.0
rootdir: /tests
plugins: timeout-2.1.0, rerunfailures-10.2, anyio-3.6.2
collected 101 items

tests/test_e2e_14_mef_eline.py x                                         [  0%]
tests/test_e2e_10_mef_eline.py ..........ss.....x.....x................  [ 40%]
tests/test_e2e_11_mef_eline.py ......                                    [ 46%]
tests/test_e2e_12_mef_eline.py .....Xx.                                  [ 54%]
tests/test_e2e_13_mef_eline.py .....xs.s......xs.s.XXxX.xxxx..X......... [ 95%]
...                                                                      [ 98%]
tests/test_e2e_15_mef_eline.py ..                                        [100%]

```

**Comment by Alopalao** [comment]
Updated changelog

**Comment by viniarck** [comment]
Let's ship it.
